### PR TITLE
DOCS-1422: Documentation for EGW destination based routing

### PR DESCRIPTION
**Pull request by mazdakn**
Documentation for destination based routing in Egress Gateways

Product Version(s):

Calico Enterprise 3.17, Next
Calico Cloud Next

Issue:

Link to docs preview:

SME review:
- [ ] An SME has approved this change. 

DOCS review:
- [ ] A member of the docs team has approved this change.

Additional information:

### CODE REVIEW

**Comment by ctauchen** [comment]
@mazdakn 

Thanks! I think it's easier to add my suggestions here:

<pre>
### Configuring namespaces and pods to use egress gateways

You can configure namespaces and pods to use an egress gateway by:
* annotating the namespace or pod
* applying an egress gateway policy to the namespace or pod.

Using an egress gateway policy is more complicated, but it allows advanced use cases.

#### Configure a namepace or pod to use an egress gateway (annotation method)

In a {{prodname}} deployment, the Kubernetes namespace and pod resources honor annotations that
tell that namespace or pod to use particular egress gateways. These annotations are selectors, and
their meaning is "the set of pods, anywhere in the cluster, that match those selectors".

So, to configure all the pods in a namespace to use the egress gateways that are
labelled with `egress-code: red`, you would annotate that namespace like this:

```bash
kubectl annotate ns <namespace> egress.projectcalico.org/selector="egress-code == 'red'"
```

By default, that selector can only match egress gateways in the same namespace. To select gateways
in a different namespace, specify a `namespaceSelector` annotation as well, like this:

```bash
kubectl annotate ns <namespace> egress.projectcalico.org/namespaceSelector="projectcalico.org/name == '<egress gateway namespace>'"
```

Egress gateway annotations have the same [syntax and range of expressions](../../reference/resources/networkpolicy.mdx#selector) as the selector fields in
{{prodname}} [network policy](../../reference/resources/networkpolicy.mdx#entityrule).

To configure a specific Kubernetes Pod to use egress gateways, specify the same annotations when
creating the pod. For example:

```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  annotations:
    egress.projectcalico.org/selector: egress-code == 'red'
    egress.projectcalico.org/namespaceSelector: projectcalico.org/name == '<egress gateway namespace>'
  name: my-client,
  namespace: <namespace>,
spec:
  ...
EOF
```

#### Configure a namespace or pod to use an egress gateway (egress gateway policy method)

Creating an egress gateway policy allows gives you more control over how your egress gateways work.
For example, you can:
  - Send egress gateway traffic to multiple egress gateways, depending on the destination.
  - Skip egress gateways for traffic that is bound for local endpoints that aren't in the cluster.

The following is an example of Egress Gateway Policy:

```yaml
apiVersion: projectcalico.org/v3
kind: EgressGatewayPolicy
metadata:
  name: "egw-policy1"
spec:
  rules:
  // callout-for-next-line
  - destination:
      cidr: 10.0.0.0/8
    description: "Local: no gateway"
  - destination:
      cidr: 11.0.0.0/8
    description: "Gateway to on prem"
    // callout-for-next-line
    gateway:
      // callout-for-next-line
      namespaceSelector: ...
      // callout-for-next-line
      selector: ...
      // callout-for-next-line
      maxNextHops: ...
  - description: "Gateway to internet"
    gateway:
      namespaceSelector: ...
      selector: ...
      maxNextHops: ...
```

<Callouts>

1. If the `destination` field is not specified, it takes the default value of 0.0.0.0/0.

1. If the `gateway` field is not specified, then egress traffic is routed locally, and not through an egress gateway.
   This is helpful for reaching local endpoints that are not part of a cluster.

1. Required when `gateway` field is specified.

1. Required when `gateway` field is specified.

1. The `maxNextHops` field specifies the maximum number of egress gateway replicas from the selected deployment that a pod depends on. For more information, see [Optimize egress networking for workloads with long-lived TCP connections](egress-gateway-maintenance.mdx#reduce-the-impact-of-gateway-downtime).

</Callouts>

{{prodname}} rejects egress gateway policies that do any of the following:
* The policy has no rule that specifies a gateway or a destination
* The policy has a rule with empty `selector` or `namespaceSelector` fields.
* The policy has two or more rules with the same destination.

To configure all the pods in a namespace to use an egress gateway policy named `egw-policy1`, you could annotate the namespace like this:

```bash
kubectl annotate ns <namespace> egress.projectcalico.org/egressGatewayPolicy="egw-policy1"
```

To configure a specific Kubernetes pod to use the same policy, specify the same annotations when
creating the pod.
For example:

```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  annotations:
    egress.projectcalico.org/egressGatewayPolicy: "egw-policy1"
  name: my-client,
  namespace: <namespace>,
spec:
  ...
EOF
```

:::caution

You must create the egress gateway policy before you apply it to a namespace or pod.
If you attempt to apply an egress gateway policy that has not been created, {{prodname}} will block all traffic from the namespace or pod.

:::

</pre>

File: calico-enterprise/networking/egress/egress-gateway-aws.mdx [hunk 1]
```diff
@@ -881,14 +889,97 @@ kind: Pod
...
+#### Configure a namespace or pod to use an egress gateway (egress gateway policy method)
+
+Creating an egress gateway policy allows gives you more control over how your egress gateways work.
+For example, you can:
+  - Send egress gateway traffic to multiple egress gateways, depending on the destination.
+  - Skip egress gateways for traffic that is bound for local endpoints that aren't in the cluster.
+
+The following is an example of Egress Gateway Policy:
+
+```yaml
```
Comment (ctauchen) [inline calico-enterprise/networking/egress/egress-gateway-aws.mdx]:
```suggestion
```yaml noValidation
```

We've set up a YAML validation test. For this sample, we need to add the noValidation tag to exclude it. Please add this to the other places as well.

**Review by ctauchen — COMMENTED** [review]
@mazdakn One small change to the ```yaml code fence syntax.

Also, we need changes to CC/Next. 

Once those are in, we should be good to go,.

File: calico-enterprise/networking/egress/egress-gateway-aws.mdx (hunk 1 above)
Comment (mazdakn) [reply calico-enterprise/networking/egress/egress-gateway-aws.mdx]: Fixed.

**Comment by mazdakn** [comment]
@ctauchen PTAL. I copied the changes to Calico Cloud, but I am not entirely sure if you meant it by referring to `Next`. Please let me know if I did it wrong.

**Pull request merged**